### PR TITLE
Feature/implement message catch and throw events

### PR DESCRIPTION
**Pull request by ElRaptorus**
## What did you change?

Implement handlers for the `MessageBoundaryEvent`, aswell as the `IntermediateMessageCatchEvent` and `IntermediateMessageThrowEvent`.

## How can others test the changes?

Run the integrationtests.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

**Comment by ElRaptorus** [comment]
https://github.com/process-engine-pro/organization/issues/182

File: src/runtime/engine/handler/intermediate_catch_event_handler.ts [hunk 1]
```diff
@@ -8,22 +8,30 @@ import {
   Runtime,
 } from '@process-engine/process_engine_contracts';
 
+import {IEventAggregator, ISubscription} from '@essential-projects/event_aggregator_contracts';
+
 import {FlowNodeHandler} from './index';
 
-export class IntermediateCatchEventHandler extends FlowNodeHandler<Model.Events.Event> {
+export class IntermediateCatchEventHandler extends FlowNodeHandler<Model.Events.IntermediateCatchEvent> {
```
Comment (roblen45) [inline src/runtime/engine/handler/intermediate_catch_event_handler.ts]: Bitte eigene Klassen für MessageIntermediateCatchEvents usw. ausprägen.

Sonst ist der ganze Code für Timer, Error, Condition und weitere in dieser Klasse.

File: src/runtime/engine/handler/intermediate_throw_event_handler.ts [hunk 2]
```diff
@@ -8,22 +8,30 @@ import {
   Runtime,
 } from '@process-engine/process_engine_contracts';
 
+import {IEventAggregator} from '@essential-projects/event_aggregator_contracts';
+
 import {FlowNodeHandler} from './index';
 
-export class IntermediateThrowEventHandler extends FlowNodeHandler<Model.Events.Event> {
+export class IntermediateThrowEventHandler extends FlowNodeHandler<Model.Events.IntermediateThrowEvent> {
```
Comment (roblen45) [inline src/runtime/engine/handler/intermediate_throw_event_handler.ts]: s.o.

File: src/runtime/engine/handler/message_boundary_event_handler.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+      const messageBoundaryEvent: Model.Events.BoundaryEvent = await this._getMessageBoundaryEvent(flowNode, processModelFacade);
+
+      let messageReceived: boolean = false;
+      let handlerHasFinished: boolean = false;
+
+      const messageName: string =
+        `/processengine/process/${token.processInstanceId}/message/${messageBoundaryEvent.messageEventDefinition.messageRef}`;
+
+      const messageReceivedCallback: any = async(): Promise<void> => {
```
Comment (roblen45) [inline src/runtime/engine/handler/message_boundary_event_handler.ts]: In eigene Methode auslagern

Comment (ElRaptorus) [reply src/runtime/engine/handler/message_boundary_event_handler.ts]: Das geht nicht, da der Callback das Promise von `executeInternally` resolven muss, sobald eine Message empfangen wurde. 
Würde man das hier nicht machen, würde die Funktion so lange weiterlaufen, bis der nächste Flow Node durchgelaufen ist, was man bei einem interrupting event ja garnicht will.

**Comment by ElRaptorus** [comment]
@roblen45 done

File: src/runtime/engine/handler/intermediate_event_handlers/intermediate_message_catch_event_handler.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                                    token: Runtime.Types.ProcessToken,
+                                    processTokenFacade: IProcessTokenFacade,
+                                    processModelFacade: IProcessModelFacade,
+                                    executionContextFacade: IExecutionContextFacade): Promise<NextFlowNodeInfo> {
+
+    const flowNodeInstanceId: string = super.createFlowNodeInstanceId();
+
+    await this.flowNodeInstanceService.persistOnEnter(executionContextFacade, token, flowNode.id, flowNodeInstanceId);
+
+    await this._waitForMessage(token.processInstanceId, flowNode.messageEventDefinition.messageRef);
```
Comment (5ebastianMeier) [inline src/runtime/engine/handler/intermediate_event_handlers/intermediate_message_catch_event_handler.ts]: Before we wait on the message, we have to suspend the `FlowNodeInstance`.

Right after we got the message, we first have to continue execution (counterpart to `suspend`) and then continue the original execution.

File: src/runtime/engine/handler/intermediate_throw_event_handler.ts [hunk 5]
```diff
@@ -10,33 +10,70 @@ import {
...
                                     processTokenFacade: IProcessTokenFacade,
                                     processModelFacade: IProcessModelFacade,
                                     executionContextFacade: IExecutionContextFacade): Promise<NextFlowNodeInfo> {
 
+    const flowNodeInstanceService: IFlowNodeInstanceService = await this.container.resolveAsync<IFlowNodeInstanceService>('FlowNodeInstanceService');
+
     const flowNodeInstanceId: string = super.createFlowNodeInstanceId();
 
-    await this.flowNodeInstanceService.persistOnEnter(executionContextFacade, token, flowNode.id, flowNodeInstanceId);
+    await flowNodeInstanceService.persistOnEnter(executionContextFacade, token, flowNode.id, flowNodeInstanceId);
```
Comment (5ebastianMeier) [inline src/runtime/engine/handler/intermediate_throw_event_handler.ts]: Same as above - `persistOnEnter` and `persistOnExit` have to happen at different times.

File: src/runtime/engine/handler/message_boundary_event_handler.ts (hunk 3 above)
Comment (5ebastianMeier) [reply src/runtime/engine/handler/message_boundary_event_handler.ts]: Doch könnte man auslagern, einfach eine Methode `createMessageReceivedCallback(messageReceivedState)` machen

Wobei `messageReceivedState` die Referenzen für die notwendigen Eigenschaften beinhaltet: `resolve, messageReceived, handlerHasFinished, ...`

Das Interface für `messageReceivedState` müsste man nichtmal exportieren, da es nur private hier in der Klasse verwendet wird

**Comment by ElRaptorus** [comment]
@5ebastianMeier done

**Pull request merged**